File: src/utils/syncBaseline.ts
import type { PersistedBoard } from '../types/persistence'
import { canonicalJson, sha256Hex, splitCloudBoard } from './cloudDocuments'

// ---------------------------------------------------------------------------
// The last-synced copy: what makes a silent reconcile possible.
//
// After every successful sync this records, per account, the exact board both
// sides agreed on. The next reconcile compares this device and the cloud to
// THAT copy instead of to each other, which is the difference between knowing
// who edited what and having to ask the user.
//
// It lives in its own IndexedDB database rather than localStorage for two
// reasons: a full board does not fit comfortably in localStorage, and a wiped
// localStorage (which is what leaves people signed out after an app update)
// must not also destroy the lineage and drag the conflict prompt back.
// ---------------------------------------------------------------------------

const DB_NAME = 'grovepad-sync'
const DB_VERSION = 1
const BASELINE_STORE = 'baselines'

export interface BoardFingerprint {
  /** Matches `board_indexes.checksum` for the same board, byte for byte. */
  indexChecksum: string
  /** canvasId -> the checksum `canvas_docs.checksum` holds for that canvas. */
  canvasChecksums: Record<string, string>
}

export interface SyncBaseline extends BoardFingerprint {
  userId: string
  /** The agreed board itself — the base side of the three-way merge. */
  board: PersistedBoard
  cloudUpdatedAt: string | null
  at: number
}

/**
 * Hash a board the way the cloud hashes it. Local and remote fingerprints are
 * therefore directly comparable, so a reconcile can decide whether anything
 * moved from two tiny metadata reads and no board transfer at all.
 */
export async function fingerprintBoard(board: PersistedBoard): Promise<BoardFingerprint> {
  const split = splitCloudBoard(board)
  const entries = await Promise.all(
    Object.entries(split.canvases).map(async ([canvasId, document]) =>
      [canvasId, await sha256Hex(canonicalJson(document))] as const),
  )
  return {
    indexChecksum: await sha256Hex(canonicalJson(split.index)),
    canvasChecksums: Object.fromEntries(entries),
  }
}

export function fingerprintsMatch(left: BoardFingerprint, right: BoardFingerprint): boolean {
  if (left.indexChecksum !== right.indexChecksum) return false
  const leftIds = Object.keys(left.canvasChecksums)
  if (leftIds.length !== Object.keys(right.canvasChecksums).length) return false
  return leftIds.every((id) => left.canvasChecksums[id] === right.canvasChecksums[id])
}

function openDatabase(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    let request: IDBOpenDBRequest
    try {
      request = indexedDB.open(DB_NAME, DB_VERSION)
    } catch (error) {
      reject(error instanceof Error ? error : new Error('Sync baseline database unavailable'))
      return
    }
    request.onerror = () => reject(request.error ?? new Error('Unable to open the sync baseline database'))
    // Another tab holding an older version open would otherwise hang this
    // promise forever, and every reconcile behind it.
    request.onblocked = () => reject(new Error('Sync baseline database upgrade is blocked'))
    request.onupgradeneeded = () => {
      const database = request.result
      if (!database.objectStoreNames.contains(BASELINE_STORE)) {
        database.createObjectStore(BASELINE_STORE, { keyPath: 'userId' })
      }
    }
    request.onsuccess = () => resolve(request.result)
  })
}

function requestResult<T>(request: IDBRequest<T>): Promise<T> {
  return new Promise((resolve, reject) => {
    request.onsuccess = () => resolve(request.result)
    request.onerror = () => reject(request.error ?? new Error('Sync baseline request failed'))
  })
}

function isBaseline(value: unknown): value is SyncBaseline {
  if (typeof value !== 'object' || value === null) return false
  const record = value as Record<string, unknown>
  return typeof record.userId === 'string' &&
    typeof record.indexChecksum === 'string' &&
    typeof record.canvasChecksums === 'object' && record.canvasChecksums !== null &&
    typeof record.board === 'object' && record.board !== null
}

/**
 * The baseline for one account, or null when there is none to reconcile
 * against. A storage failure reads as "no lineage" rather than throwing: the
 * merge then unions both sides, which is lossless, instead of blocking sync.
 */
export async function readSyncBaseline(userId: string): Promise<SyncBaseline | null> {
  try {
    const database = await openDatabase()
    try {
      const stored = await requestResult(
        database.transaction(BASELINE_STORE).objectStore(BASELINE_STORE).get(userId),
      )
      return isBaseline(stored) && stored.userId === userId ? stored : null
    } finally {
      database.close()
    }
  } catch {
    return null
  }
}

/** Record the agreed board. Best-effort: a failure only costs the next merge
 * its lineage, so it must never fail a sync that already succeeded. */
export async function writeSyncBaseline(baseline: SyncBaseline): Promise<void> {
  try {
    const database = await openDatabase()
    try {
      const transaction = database.transaction(BASELINE_STORE, 'readwrite')
      transaction.objectStore(BASELINE_STORE).put(baseline)
      await new Promise<void>((resolve, reject) => {
        transaction.oncomplete = () => resolve()
        transaction.onerror = () => reject(transaction.error ?? new Error('Baseline write failed'))
        transaction.onabort = () => reject(transaction.error ?? new Error('Baseline write aborted'))
      })
    } finally {
      database.close()
    }
  } catch {
    // Nothing to recover: the next reconcile simply merges without lineage.
  }
}

File: apple/Sources/GrovepadCore/Sync/SyncBaseline.swift
import Foundation

// ---------------------------------------------------------------------------
// Port of `src/utils/syncBaseline.ts`: the last-synced copy, what makes a
// silent reconcile possible.
//
// After every successful sync this records, per account, the exact board both
// sides agreed on. The next reconcile compares this device and the cloud to
// THAT copy instead of to each other, which is the difference between knowing
// who edited what and having to ask the user.
//
// The web keeps it in its own IndexedDB database so a wiped localStorage
// cannot destroy the lineage. Here it is its own file under the store
// directory, one per account: `<store>/sync/baselines/<account>.json`.
// `fingerprintBoard` itself lives in `Persistence/CloudDocuments.swift`.
// ---------------------------------------------------------------------------

public struct SyncBaseline: Equatable {
    public var userId: String
    /// The agreed board itself — the base side of the three-way merge — as a
    /// serialized persisted board document.
    public var board: JSONObject
    /// Matches `board_indexes.checksum` for the same board, byte for byte.
    public var indexChecksum: String
    /// canvasId -> the checksum `canvas_docs.checksum` holds for that canvas.
    public var canvasChecksums: OrderedMap<String>
    public var cloudUpdatedAt: String?
    /// Milliseconds since the epoch, as `Date.now()` writes it.
    public var at: Double

    public init(userId: String, board: JSONObject, fingerprint: CloudDocuments.BoardFingerprint, cloudUpdatedAt: String?, at: Double) {
        self.userId = userId
        self.board = board
        self.indexChecksum = fingerprint.indexChecksum
        self.canvasChecksums = fingerprint.canvasChecksums
        self.cloudUpdatedAt = cloudUpdatedAt
        self.at = at
    }

    public var fingerprint: CloudDocuments.BoardFingerprint {
        CloudDocuments.BoardFingerprint(indexChecksum: indexChecksum, canvasChecksums: canvasChecksums)
    }

    // MARK: - Codec

    /// The stored record, in the web's key order.
    public func serialized() -> JSONObject {
        var object = JSONObject()
        object["userId"] = .string(userId)
        object["board"] = .object(board)
        object["indexChecksum"] = .string(indexChecksum)
        var checksums = JSONObject()
        for (canvasId, checksum) in canvasChecksums.entries { checksums[canvasId] = .string(checksum) }
        object["canvasChecksums"] = .object(checksums)
        object["cloudUpdatedAt"] = cloudUpdatedAt.map { .string($0) } ?? .null
        object["at"] = .number(at)
        return object
    }

    /// `isBaseline`: a stored value is trusted only with every required field
    /// in the right shape; anything else reads as "no lineage".
    public static func parse(_ value: JSONValue) -> SyncBaseline? {
        guard let object = value.objectValue,
              let userId = object.string("userId"), object.isString("userId"),
              let indexChecksum = object.string("indexChecksum"), object.isString("indexChecksum"),
              let checksums = object.object("canvasChecksums"),
              let board = object.object("board") else { return nil }
        var canvasChecksums = OrderedMap<String>()
        for (canvasId, checksum) in checksums.entries {
            guard let text = checksum.stringValue else { continue }
            canvasChecksums[canvasId] = text
        }
        return SyncBaseline(
            userId: userId,
            board: board,
            fingerprint: CloudDocuments.BoardFingerprint(indexChecksum: indexChecksum, canvasChecksums: canvasChecksums),
            cloudUpdatedAt: object.string("cloudUpdatedAt"),
            at: object.number("at") ?? 0
        )
    }
}

/// Reads and writes one baseline file per account. Every failure reads as
/// "no lineage" and every write is best-effort: a failure only costs the next
/// merge its baseline (it then unions, which is lossless) and must never fail
/// a sync that already succeeded.
public final class SyncBaselineStore {
    public let directory: URL

    /// - Parameter storeDirectory: the `LocalBoardStore` directory; baselines
    ///   live beside the board in `sync/baselines/`.
    public init(storeDirectory: URL) {
        directory = storeDirectory
            .appendingPathComponent("sync", isDirectory: true)
            .appendingPathComponent("baselines", isDirectory: true)
    }

    public func fileURL(userId: String) -> URL {
        directory.appendingPathComponent("\(Self.fileName(userId)).json")
    }

    /// Account ids are UUIDs, which are file-name safe; anything else is
    /// hashed so a stray separator can never escape the directory.
    public static func fileName(_ userId: String) -> String {
        let safe = userId.unicodeScalars.allSatisfy { scalar in
            ("a"..."z").contains(scalar) || ("A"..."Z").contains(scalar) || ("0"..."9").contains(scalar) || scalar == "-" || scalar == "_"
        }
        return safe && !userId.isEmpty ? userId : "h-" + SHA256.hex(userId)
    }

    /// `readSyncBaseline`: the baseline for one account, or nil when there is
    /// none to reconcile against (or it belongs to somebody else, or it cannot
    /// be read).
    public func read(userId: String) -> SyncBaseline? {
        guard let data = try? Data(contentsOf: fileURL(userId: userId)),
              let value = try? JSONParser.parse(data),
              let baseline = SyncBaseline.parse(value),
              baseline.userId == userId else { return nil }
        return baseline
    }

    /// `writeSyncBaseline`: record the agreed board. Never throws.
    public func write(_ baseline: SyncBaseline) {
        do {
            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
            try Data(JSONWriter.stringify(.object(baseline.serialized())).utf8)
                .write(to: fileURL(userId: baseline.userId), options: .atomic)
        } catch {
            // Nothing to recover: the next reconcile simply merges without lineage.
        }
    }

    /// Forget one account's lineage (sign-out teardown).
    public func clear(userId: String) {
        try? FileManager.default.removeItem(at: fileURL(userId: userId))
    }

    /// Forget every account's lineage.
    public func clearAll() {
        try? FileManager.default.removeItem(at: directory)
    }
}
